### PR TITLE
BUG : cascaded filters in image objects

### LINKED ISSUE #1912: UnboundLocalError: cannot access local variable 'img' where it is not associated with a value
An UnboundLocalError is thrown when I try to extract (inlined?) images

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
Linux-5.4.0-152-generic-x86_64-with-glibc2.31

$ python -c "import pypdf;print(pypdf.__version__)"
3.11.0
```

## Code + PDF

This is a minimal, complete example that shows the issue:

```python
from pypdf import PdfReader
import pypdf

reader = PdfReader("GeoTopo-komprimiert.pdf")
for page in reader.pages:
    for image in page.images:
        print(image.name)
```

The file: https://github.com/py-pdf/pypdf/files/11845099/GeoTopo-komprimiert.pdf

## Traceback

```
Traceback (most recent call last):
  File "/home/moose/Downloads/size/code.py", line 7, in <module>
    for image in page.images:
  File "/home/moose/Github/py-pdf/pypdf/pypdf/_page.py", line 2555, in __iter__
    yield self[i]
          ~~~~^^^
  File "/home/moose/Github/py-pdf/pypdf/pypdf/_page.py", line 2551, in __getitem__
    return self.get_function(lst[index])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/moose/Github/py-pdf/pypdf/pypdf/_page.py", line 522, in _get_image
    imgd = _xobj_to_image(cast(DictionaryObject, xobjs[id]))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/moose/Github/py-pdf/pypdf/pypdf/filters.py", line 852, in _xobj_to_image
    alpha = _xobj_to_image(x_object_obj[IA.S_MASK])[2]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/moose/Github/py-pdf/pypdf/pypdf/filters.py", line 879, in _xobj_to_image
    img.save(img_byte_arr, format=image_format)
    ^^^
UnboundLocalError: cannot access local variable 'img' where it is not associated with a value
```

### PR DESCRIPTION
**Pull request by pubpub-zz**
closes #1912
the issue is dealing with some images being encoded with 2 filters being cascaded

### CODE REVIEW

**Comment by pubpub-zz** [comment]
Do not have test data for the moment to improve coverage
@MartinThoma 
All yours :)

File: pypdf/filters.py [hunk 1]
```diff
@@ -742,108 +742,97 @@ def _xobj_to_image(x_object_obj: Dict[str, Any]) -> Tuple[Optional[str], bytes,
...
-                            "1": (0, "", ""),
-                            "L": (1, "P", "L"),
-                            "P": (0, "", ""),
-                            "RGB": (3, "P", "RGB"),
-                            "CMYK": (4, "P", "CMYK"),
-                        }[_get_imagemode(base, 0, "")]
-                    except KeyError:  # pragma: no cover
+    filters = x_object_obj.get(SA.FILTER, [None])
+    lfilters = filters[-1] if isinstance(filters, list) else filters
+    if lfilters == FT.FLATE_DECODE:
```
Comment (MartinThoma) [inline pypdf/filters.py:747]: I think the code might be more readable if we had some extra functions. Something like this:

```python
if lfilters == FT.FLATE_DECODE:
    img, image_format, extension = _handle_flate(size, data, mode, color_space)
elif lfilters in (FT.LZW_DECODE, FT.ASCII_85_DECODE, FT.CCITT_FAX_DECODE):
    # Code block: Maybe leave that as-is
elif lfilters == FT.JPX_DECODE:
    img, image_format, extension = _handle_flate(size, data, mode, color_space)
elif lfilters == FT.CCITT_FAX_DECODE:
    # Code block: Maybe leave that also as is
elif lfilters is None:
    # Code block: Maybe leave that also as is
```

**Comment by MartinThoma** [comment]
I will have another look at it tomorrow (I want to try the extraction + inspect the files manually). But I'm very certain that I'm taking this into pypdf tomorrow + realease it :-)

**Comment by MartinThoma** [comment]
I can confirm that the extraction to jp2 worked. I needed to follow https://askubuntu.com/a/1151781 to convert it to something I can view though xD

Extracting `007-imagemagick-images/imagemagick-ASCII85Decode.pdf` failed:

> UnidentifiedImageError: cannot identify image file

**Pull request merged**